NPE from instructorCoursesPage

java.lang.NullPointerException

Request Path /page
Request Parameters {next::/page/instructorCoursesPage, user::abc, courseid::GEOL211_SP2014}
Stack Trace

java.lang.RuntimeException: java.lang.NullPointerException
    at teammates.common.util.StringHelper.encrypt(StringHelper.java:87)
    at teammates.storage.api.InstructorsDb.deleteDocument(InstructorsDb.java:57)
    at teammates.storage.api.InstructorsDb.deleteInstructorsForCourse(InstructorsDb.java:417)
    at teammates.logic.core.InstructorsLogic.deleteInstructorsForCourse(InstructorsLogic.java:377)
    at teammates.logic.core.CoursesLogic.deleteCourseCascade(CoursesLogic.java:643)
    at teammates.logic.api.Logic.deleteCourse(Logic.java:844)
    at teammates.ui.controller.InstructorCourseDeleteAction.execute(InstructorCourseDeleteAction.java:25)
    at teammates.ui.controller.Action.executeAndPostProcess(Action.java:312)
    at teammates.ui.controller.ControllerServlet.doPost(ControllerServlet.java:58)
    at teammates.ui.controller.ControllerServlet.doGet(ControllerServlet.java:37)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:617)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
    at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
    at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
    at teammates.storage.datastore.DatastoreFilter.doFilter(DatastoreFilter.java:28)
    at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
    at com.google.appengine.tools.appstats.AppstatsFilter.doFilter(AppstatsFilter.java:142)
    at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
    at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
    at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
    at com.google.apphosting.runtime.jetty.SaveSessionFilter.doFilter(SaveSessionFilter.java:35)
    at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
    at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
    at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
    at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:43)
    at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
    at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
    at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
    at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
    at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
    at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
    at com.google.apphosting.runtime.jetty.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:254)
    at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
    at org.mortbay.jetty.Server.handle(Server.java:326)
    at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
    at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
    at com.google.apphosting.runtime.jetty.RpcRequestParser.parseAvailable(RpcRequestParser.java:76)
    at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
    at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:146)
    at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:446)
    at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:438)
    at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:445)
    at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:220)
    at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:309)
    at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:301)
    at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:442)
    at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:251)
    at java.lang.Thread.run(Thread.java:724)
Caused by: java.lang.NullPointerException
    at teammates.common.util.StringHelper.encrypt(StringHelper.java:84)
    ... 47 more

--- a/src/main/java/teammates/storage/api/InstructorsDb.java
+++ b/src/main/java/teammates/storage/api/InstructorsDb.java
@@ -48,13 +48,20 @@ public class InstructorsDb extends EntitiesDb{
         if(instructor.key == null){
             instructor = this.getInstructorForEmail(instructor.courseId, instructor.email);
         }
-        putDocument(Const.SearchIndex.INSTRUCTOR, new InstructorSearchDocument(instructor));
+        // defensive coding for legacy data
+        if(instructor.key != null) {
+            putDocument(Const.SearchIndex.INSTRUCTOR, new InstructorSearchDocument(instructor));
+        }
     }
     
     public void deleteDocument(InstructorAttributes instructorToDelete){
         if(instructorToDelete.key == null){
             InstructorAttributes instructor = this.getInstructorForEmail(instructorToDelete.courseId, instructorToDelete.email);
-            deleteDocument(Const.SearchIndex.INSTRUCTOR, StringHelper.encrypt(instructor.key));
+            
+            // handle legacy data which do not have key attribute (key == null)
+            if(instructor.key != null) {
+                deleteDocument(Const.SearchIndex.INSTRUCTOR, StringHelper.encrypt(instructor.key));
+            }
         } else {
             deleteDocument(Const.SearchIndex.INSTRUCTOR, StringHelper.encrypt(instructorToDelete.key));
         }

